### ISSUE TITLE
Support tsconfig.json paths and baseUrl automatically

### ISSUE THREAD

**Issue by lydell** [lydell/eslint-plugin-simple-import-sort#31]
It would be cool if the plugin could automatically detect tsconfig.json to see which first party imports you have, such as `web/components/foo`. Then those could be moved into the “absolute imports” group, instead of ending up in the (third party) “packages” group.

One idea is to use https://github.com/dividab/tsconfig-paths. Need to check that it actually supports baseUrl, though.

Another idea is to try to `require("typescript")`. If it’s there, use it to resolve the config.

The initial idea comes from: https://github.com/lydell/eslint-plugin-simple-import-sort/issues/18#issuecomment-532097832

**Comment by zbeyens** [comment]
You can also take a look to [import/order](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/order.md) that can differentiate external and internal modules. 

```
// 2. "external" modules
import _ from 'lodash';
import chalk from 'chalk';
// 3. "internal" modules
// (if you have configured your path or webpack to handle your internal paths differently)
import foo from 'src/foo';
```

**Comment by anilanar** [comment]
I had terrible experience (in terms of performance/caching) so far with plugins/rules that relies on external "effects" (config files, files other than the file being linted etc.).
I've never implemented a plugin myself, and I always keep TS rules that rely on compiler features out.
I'd feel very sorry if the same happened to this awesome tool.

**Comment by lydell** [comment]
@anilanar Thanks for sharing your experience! That’ll be kept in mind if this is implemented. I think the added [custom grouping](https://github.com/lydell/eslint-plugin-simple-import-sort#custom-grouping) option works well enough and is easy enough to use that this might not even be needed.

**Comment by israelKusayev** [comment]
any progress with `baseUrl` support 👀?

**Comment by lydell** [comment]
No, I’m not so interested in this anymore because (1) I don’t need this myself and (2) the custom grouping option seems easy enough to set up (and is possibly more performant).

Would you like to tell us more about your use case that you’d want this for and how much configuration you’d avoid not having to set up custom grouping?

**Comment by israelKusayev** [comment]
I have a create-react-app typescript project, and I've set the project to use non-relative modules.
my tsconfig:
```
{
  "compilerOptions": {
    "target": "es5",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "jsx": "react",
    "rootDirs": ["src", "stories"],
    "baseUrl": "src" 
  },
  "include": ["src", "stories"]
}
```

so now I can use 

```javascript
import CommonService from 'services/commonService';
```
insted of
```javascript
import CommonService from '../../../services/commonService';
```

and now when I'm using the auto-fix import
my imports look like that
```javascript
import { Col, Row } from 'antd'; // third party
import Dictionary from 'dictionary/dictionary'; // my module
import React, { useState } from 'react'; // third party
import { RouteComponentProps } from 'react-router-dom'; // third party
import CommonService from 'services/commonService'; // my module
import TreatmentService from 'services/treatmentService';// my module
```

which makes no sense

**Comment by lydell** [comment]
@israelKusayev Thanks! Have you tried the custom grouping option? How did it go?

**Comment by tpict** [comment]
I don't know that custom grouping is the right solution here–won't it require you to write a regex pattern that explicitly matches every file/directory in your `baseUrl`?

**Comment by lydell** [comment]
Yes it does! By showing me how complex that would be, that would motivate adding the automatic tsconfig.json thing.

**Comment by israelKusayev** [comment]
Sorry for the late response

I tried to use groups and I ended up with:
```javascript
"overrides": [
    {
      "files": ["*.tsx", "*.ts"],
      "rules": {
        "simple-import-sort/sort": [
          "error",
          {
            "groups": [
              // Packages. `react` related packages come first.
              // Things that start with a letter (or digit or underscore), or `@` followed by a letter.
              ["^react", "^@?\\w"],
              // Absolute imports and Relative imports.
              ["^(utils|services|hooks|hoc|types|contexts|dictionary|components)(/.*|$)", "^\\."],
              // for scss imports.
              ["^[^.]"]
            ]
          }
        ]
      }
    }
  ]
```

I think it is great for me, except for one thing, If I'll add more folders to my project under `src` folder, I'll need to update eslint file 

But I'm not adding a new folder (under `src`) every day, so it is OK (i think..)

**Comment by lydell** [comment]
@israelKusayev Thanks!

If you use .eslintrc.<strong>js</strong>, maybe you can use `fs.readDirSync("src")` if you don’t want to update the config every time you add a folder to `src/`?

**Comment by israelKusayev** [comment]
Yesss great idea, works perfect
```javascript
const fs = require('fs');

const foldersUnderSrc = fs
  .readdirSync('src', { withFileTypes: true })
  .filter(dirent => dirent.isDirectory())
  .map(dirent => dirent.name);
```

and then 
```javascript
// Absolute imports and Relative imports.
[`^(${folders.join('|')})(/.*|$)`, '^\\.'],
```
You can see here the config file https://github.com/reflexology/client-V2/blob/master/.eslintrc.js

**Comment by lydell** [comment]
Cool! If you’re paranoid you could regex-escape the directory names as well :)

**Comment by iwan933** [comment]
The import plugin solves this using custom resolvers, which works well with `eslint-import-resolver-typescript`. See the issue in their repository [here](https://github.com/benmosher/eslint-plugin-import/issues/1485). Could be a reasonable extension here to omit double configuration.

**Comment by lydell** [comment]
@iwan933 Can you show what your double configuration looks like?

**Comment by iwan933** [comment]
@lydell double configuration is maybe the wrong wording here. But actually if i would use the simple-import-sort and import plugin, it would be nice to have a single solution for handling tsconfigs baseUrl. Personally i prefer the resolver over the groups solution. But i might miss some points here as i have not developed a plugin yet.

**Comment by lydell** [comment]
Ok! It still helps to see some real config, though. Having examples to look at always help me.

**Comment by JounQin** [comment]
Can we reuse `resolver` plugins of `eslint-plugin-import`?

**Comment by lydell** [comment]
@JounQin Last time I looked at the resolvers of eslint-plugin-import I came to the conslusion that we can’t reuse them. Things might have changed though and I might have overlooked something, so feel free to look into it!

But do keep in mind that:

- I would like to see a really compelling real-world use case before we add complexity to this plugin.
- I really like how this plugin is very small and has no dependencies. My bar is pretty high when it comes to changes to that. So don’t spend lots of time on a PR or something without talking about it first, or not knowing that it might not be merged.

**Comment by lydell** [comment]
> The resolvers are installed by users manually, so this is not a problem?

We clearly have different opinions here. I don’t think anything good will come out of a discussion about it.

**Comment by JounQin** [comment]
`require("typescript")` is not even used in [`eslint-import-resolver-typescript`](https://github.com/alexgorbatchev/eslint-import-resolver-typescript/blob/master/src/index.ts#L45), it is using [`tsconfig-paths`](https://github.com/dividab/tsconfig-paths) inside.

I was thinking to reuse `import/resolver`'s settings, so I don't know what you want here.

**Comment by lydell** [comment]
Ok.

**Comment by lydell** [comment]
I’ve made up my mind. This is complexity I don’t want. I’m updating the readme: #93
